modules/storage/cinder.go

```diff
@@ -0,0 +1,76 @@
+/*
+Copyright 2022 Red Hat
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package storage
+
+import (
+	corev1 "k8s.io/api/core/v1"
+)
+
+const (
+	// CinderExtraVolTypeUndefined can be used to label an extraMount which
+	// is not associated with a specific backend
+	CinderExtraVolTypeUndefined ExtraVolType = "Undefined"
+	// CinderExtraVolTypeCeph can be used to label an extraMount which
+	// is associated to a Ceph backend
+	CinderExtraVolTypeCeph      ExtraVolType = "Ceph"
+
+	// CinderBackup is the definition of the cinder-backup group
+	CinderBackup    ServiceType = "CinderBackup"
+	// CinderVolume is the definition of the cinder-volume group
+	CinderVolume    ServiceType = "CinderVolume"
+	// CinderScheduler is the definition of the cinder-scheduler group
+	CinderScheduler ServiceType = "CinderScheduler"
+	// CinderAPI is the definition of the cinder-api group
+	CinderAPI       ServiceType = "CinderAPI"
+
+	// Cinder is the global ServiceType that refers to all the components deployed
+	// by the cinder operator
+	Cinder ServiceType = "Cinder"
+)
+
+// CinderExtraVolMounts exposes additional parameters processed by the cinder-operator
+// and defines the common VolMounts structure provided by the main storage module
+type CinderExtraVolMounts struct {
+	// +kubebuilder:validation:Optional
+	Name string `json:"name,omitempty"`
+	// +kubebuilder:validation:Optional
+	Region string `json:"region,omitempty"`
+	// +kubebuilder:validation:Required
+	VolMounts []VolMounts `json:"extraVol,omitempty"`
+}
```

Comment:
It looks like this struct is the same across `Cinder`, `Glance` and `Openstack`.  Is there a reason to not just combine them as a single struct type?  Perhaps you're envisioning more fields in the future that might be needed for one service but not another?

Reply:
It's the same today but we can imagine them differing in the future per type so that's why each service has its own version.

Reply:
Yeah, it's true that for now these data structures are pretty much the same, and
the same thing applies to the functions related to cinder and glance.
However, in my vision I see two main points that led me to keep them separated:

1. the data structures can diverge at some point (same thing for the associated
functions); glance might add parameters to handle more granular policies while cinder
can improve the `ExtraVolType` to mount only **"trusted"** volumes; this is just an example
of two implementations that can go in a different direction;

2. my idea is to minimize the changes required to the operators in case something
   change in lib-common: if I collapse everything in a single data structure, a
   PR that improves `Glance` can break `Cinder`, or just have side effects to that
   (or any other) component that is happy with the current status of the module.
   In general `GetVolumes` and `GetVolumeMounts` are common functions for all the
   operators (so that can stay in storage.go), and all the services can "extend"
   this model and build their own logic on top of it.
   The `ansible-operator` doesn't need the extra logic provided for cinder and glance,
   and it can just rely on the basic storage feature [0], hence no need to build
   anything on top of it.

With this change I'm also trying to establish a pattern that should be
followed in future developments:

- `storage.go, utils.go` -> common functions implemented for the whole storage module
- `<operator>.go` -> it's available in the storage module namespace, and it inherits the
                              basic features specializing the behavior for a given service

By doing this I can improve my service within the storage module w/o affecting any other
component that relies to this module, but, at the same time, I can build on top of the basic
building block provided by storage.go.

[0]  https://github.com/fmount/ansibleee-operator/blob/extra_volumes/api/v1alpha1/ansibleee_types.go#L54

Reply:
Two questions
* Do I get it right that these operator type specific types are only used in those operators? like Cinder in cinder-operator, Glance in glance-operator, ... . If so, shouldn't those be part of the operator and the struct be part of their versioned api? * From what you said, you expect those might change in future, so I think we might want to version them like an operator api, as a module version is probably not enough? Having those in the operator itself might not required that?

Reply:
Fair point, and you're right about thinking that this should be versioned as long as the structure evolves; we can probably reduce the maintenance over lib-common instead of growing it: I'm just not sure this struct will change too much over the time, and considering that the operators define the same structure on each service that can use it, the idea is to reduce redundancy and avoid messing up w/ the operator flow, but maybe this "versioning" work can be part of a follow up change once the feature is introduced and stable enough?
For now I know that any code change can go in a single place in lib-common, which is easier to maintain, but I can definitely experiment with it and propose a follow up as long as we have the big picture in mind.

Reply:
I agree that it the struct likely does not change too much and I like the approach. I just want to make us think about how we handle changes to the struct for the future. we need to be able to do auto api version transitions of the operator using this module and expose the ExtraMounts. since all methods/func depend on the structs, I think we could just move all to a v1beta1 package to introduce the version and keep it forever if there is no next version required.

Reply:
Right, that's a good point we were discussing w/ @gibizer in other comments.
I think we have a couple of choices here:

1. versioning the lib-common module (as you mentioned in your previous comment with the v1beta package part)
2. Moving the operators specific implementation in the operators itself

The advantage of doing .2 is that we can version a `CinderExtraVolMounts` structure within `cinder/api/v1beta/cinder_types.go`,  but we're splitting the complexity between lib-common and cinder-operator, which logically makes sense, just not sure if it's be best path we can take.
I think in general the versioning is a problem in all  lib-common modules: if you break keystone you end up breaking all the operators that use it, unless you explicitly  downgrade lib-common in go.mod: that's why I would just think more about the problem and propose a pattern that solves this potential problem globally.

Reply:
I think it would be better not having the `<operator>.go` modules here, as it seems conceptually wrong to have a cinder specific structure residing outside of the cinder-operator.

Reply:
ack, and actually, as we were discussing w/ @stuggi we should have them in the operator api/ section, so that this structure can be properly versioned: the advantage is to avoid a change in lib-common that can potentially break the operators code.
I'll try this in my next iteration in a bit.

Reply:
@Akrog @stuggi  in the last update here I removed the `<operator>.go` data structures which are now versioned as part of the operator(s) API (see [1][2][3]).
I guess we can now test all this part and move forward with the lib-common change.

[1] https://github.com/openstack-k8s-operators/cinder-operator/pull/65/files#diff-4c0f5ae542ebffc716c3eb8a28926941a358007b4bac2a3efd262c173578685aR198
[2] https://github.com/openstack-k8s-operators/glance-operator/pull/75/files#diff-f462211fddba68b6fd25298dffed9956b8b2613a6cf7d3e55f95e59a596fc0c7R191
[3] https://github.com/openstack-k8s-operators/openstack-operator/pull/38/files#diff-e003a80b7e277d733db592a080dfad3eaa21c77b28dc5df577166743b54c898bR178